pkg/virt-controller/watch/vmi.go

```diff
@@ -778,6 +791,29 @@ func (c *VMIController) updateStatus(vmi *virtv1.VirtualMachineInstance, pod *k8
 	return nil
 }
 
+func isVirtHandlerAlive(clientset kubecli.KubevirtClient, nodeName string) (bool, error) {
+	handlerNodeSelector := fields.ParseSelectorOrDie("spec.nodeName=" + nodeName)
+	handlerLabelSelector := labels.Set{virtv1.AppLabel: "virt-handler"}.AsSelector()
+	list, err := clientset.CoreV1().Pods(v1.NamespaceAll).List(context.Background(), v1.ListOptions{
+		FieldSelector: handlerNodeSelector.String(),
+		LabelSelector: handlerLabelSelector.String(),
+	})
+
+	if err != nil {
+		return false, fmt.Errorf("failed to get virt-handler pods %v", err)
+	}
+
+	if list == nil || len(list.Items) < 1 {
+		return false, fmt.Errorf("virt-handler pod not found")
+	}
+
+	for _, pod := range list.Items {
+		if podIsDown(&pod) || isHandlerContainerFailed(&pod) || pod.DeletionTimestamp != nil {
+			return false, fmt.Errorf("virt-handler is going down, node %s pod %s", nodeName, pod.Name)
+		}
+	}
```

Comment:
It might be better to reverse this function and look for a running virt-handler in order to determine if virt-handler is alive rather than looking for a dead virt-handler.

for example, consider updates here. It's possible during a rolling update that we'll have a virt-handler marked for deletion `DeletionTimestamp != nil` and a new virt-handler pod that is active.

Reply:
You are absolutely right. Thanks! Done.